streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -473,6 +474,10 @@ public static StreamThread create(final TopologyMetadata topologyMetadata,
         taskManager.setMainConsumer(mainConsumer);
         referenceContainer.mainConsumer = mainConsumer;
 
+        final String stateUpdaterId = stateUpdaterEnabled ? threadId.replace("-StreamThread-", "-StateUpdater-") : "NA";
+        final StreamsThreadDelegatingMetricsReporter reporter = new StreamsThreadDelegatingMetricsReporter(mainConsumer, threadId, stateUpdaterId);
```

Comment:
Do we need `threadId` specific `StreamsThreadDelegatingMetricsReporter` or it can be singleton?

Reply:
The short answer is that we need a `StreamsThreadDelegatingMetricsReporter` per stream thread. 

For more context, Kafka Streams has an embedded consumer for each `StreamThread` instance. Since we want each embedded consumer to register/deregister metrics only for its "host" StreamThread, we need a reporter instance per thread.